Generate a hash function when `frozen` is `True`

rhuille opened the pull request:
## Change Summary

**Before:**

```
from pydantic import BaseModel

class A(BaseModel):
    x: int
    
    class Config:
        allow_mutation = False

a = A(x=1)
d = {a: 2}

>>> TypeError: unhashable type: 'A'
```

**After:**

```py
from pydantic import BaseModel

class A(BaseModel):
    x: int
    
    class Config:
        frozen = True

a = A(x=1)
d = {a: 2}
d[a]
>>> 2
```

Note that we still have:

```py
from pydantic import BaseModel

class A(BaseModel):
    x: int
    
    class Config():
        frozen = True

a = A(x=1)
d = {a: 2}
d[a]
>>> TypeError: unhashable type: 'A'
```

Screenshot of the updated documentation:

![image](https://user-images.githubusercontent.com/15191323/106026148-bd90ab80-60c9-11eb-98ee-8a406ace4fcb.png)



## Related issue number
closes #1880

linked issue: #1303

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

ccharlesgb commented:
I think this is a good idea but I don't think this will work if the fields in the model cannot be hashed. For example what happens if you do the following:

```python
from pydantic import BaseModel

class A(BaseModel):
    x: int
    y: List[int]
    
    
    class Config:
        allow_mutation = False

a = A(x=1, y=[1,2,3])
d = {a: 2}
d[a]
>>> ?
```

Might be worth putting in a validation step to ensure that the fields are hashable?

Or alternatively `dataclasses` achieves this by assigning a `hash` attribute to each field and the overall model's hash is the combination of al of these. See: https://github.com/ericvsmith/dataclasses/blob/master/dataclasses.py#L540

I think this would be a better approach as it should allow you to do nested immutability:

```python
from pydantic import BaseModel

class B(BaseModel):
    z: int
    class Config:
        allow_mutation = False


class A(BaseModel):
    x: int
    y: B
    
    
    class Config:
        allow_mutation = False

a = A(x=1, y=B(z=4))
d = {a: 2}
```

layday commented:
I've had to roll my own `__hash__` function for Pydantic models in the past, so I'm very excited at the prospect of natively 'frozen' models.  A few quick notes - you cannot hash `self.__dict__.values()` directly because the `dict_values` container is unique.  This is why `dataclasses` places them in a tuple before hashing.  You also cannot (should not) implement a `__hash__` which behaves differently from `__eq__`:

```
In [1]: from pydantic import BaseModel

In [2]: class Foo(BaseModel):
   ...:     def __hash__(self):
   ...:         return hash(self.__dict__.values())
   ...:

In [3]: Foo() == Foo()
Out[3]: True

In [4]: hash(Foo())
Out[4]: 276660083

In [5]: hash(Foo())
Out[5]: 277242134

In [6]: d = {Foo(): True}

In [7]: d[Foo()] = False

In [8]: d.keys()
Out[8]: dict_keys([Foo(), Foo()])
```

The `dataclasses` implementation mentioned above is more sophisticated (e.g. it compares values which cannot be hashed) and might be worth investigating.

rhuille commented:
Hi @ccharlesgb and @layday , thanks for your answer.

I agree with your comments and I corrected my implementation of `__hash__` in: https://github.com/samuelcolvin/pydantic/pull/1881/commits/b1a84dc30091bc94f5457d5db0451885f0d7b7dd

What do you think ?

If I am correct, this has now the same behavior has in `dataclass`:

```
from dataclasses import dataclass
from pydantic import BaseModel
from typing import List


class A(BaseModel):
    x: int
    y: int
    
    class Config:
        allow_mutation = False


@dataclass(frozen=True)
class B:
    x: int
    y: int
 
hash(A(x=1, y=2)) == hash(B(x=1, y=2))
>>> True

hash(A(x=1, y=2)) == hash((1, 2))
>>> True

class C(BaseModel):
    x: int
    y: List[int]
    class Config:
        allow_mutation = False

@dataclass(frozen=True)
class D:
    x: int
    y: List[int]

hash(C(x=1, y=[1, 2, 3]))
>>>TypeError: unhashable type: 'list'

hash(D(x=1, y=[1, 2, 3]))
>>>TypeError: unhashable type: 'list'
```

wozniakty commented on pydantic/main.py:
```diff
@@ -313,6 +313,7 @@ def __new__(mcs, name, bases, namespace, **kwargs):  # noqa C901
             '__schema_cache__': {},
             '__json_encoder__': staticmethod(json_encoder),
             '__custom_root_type__': _custom_root_type,
+            '__hash__': (lambda self: hash(tuple(self.__dict__.values()))) if not config.allow_mutation else None,
```
Instead of `.values` we should probably have `.items` here.  The keys ought to be considered a part of the identity of the object.

Example:
```
x = {"a": 1, "b": True}
y = {"d": 1, "c": True}
hash(tuple(x.values())) == hash(tuple(y.values()))  # True
hash(tuple(x.items())) == hash(tuple(y.items()))  # False
```

rhuille replied:
In that case, we diverge from behaviour of the built-in dataclass. But I do not know if it is a problem.
If it not a problem, then I agree with you ! :)

samuelcolvin replied:
I don't think this needs to be a lambda, it can be proper method on `BaseModel` I imagine. It could raise an error if `allow_mutation is True` .

layday replied:
> I don't think this needs to be a lambda, it can be proper method on `BaseModel` I imagine. It could raise an error if `allow_mutation is True`.

The 'correct' way to make an object unhashable is by setting its `__hash__` attribute to `None`:

```
In [1]: class Foo:
   ...:     pass
   ...:

In [2]: class Bar:
   ...:     __hash__ = None
   ...:

In [3]: hash(Foo())
Out[3]: 284941123

In [4]: hash(Bar())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-d413494abdfb> in <module>
----> 1 hash(Bar())

TypeError: unhashable type: 'Bar'
```

This would have to happen on the metaclass and is done automatically by the Python interpreter for classes which implement `__eq__` but not `__hash__`:

```
In [5]: from pydantic import BaseModel

In [6]: class Baz(BaseModel):
   ...:     pass
   ...:

In [7]: hash(Baz())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-67f7cb492b68> in <module>
----> 1 hash(Baz())

TypeError: unhashable type: 'Baz'

In [8]: Baz.__hash__

In [9]: class Bee:
   ...:     def __eq__(self, other):
   ...:         ...
   ...:

In [10]: Bee.__hash__

In [11]: Foo.__hash__
Out[11]: <slot wrapper '__hash__' of 'object' objects>
```

If `__hash__` were implemented on `BaseModel`, we'd have to emulate the default behaviour.  In addition,  if `__hash__` is not `None`, `isinstance(obj, typing.Hashable)` will not have the desired effect:

```
In [12]: from typing import Hashable

In [13]: isinstance(Bee(), Hashable)
Out[13]: False

In [14]: class FauxUnhashable:
    ...:     def __hash__(self):
    ...:         raise TypeError
    ...:

In [51]: isinstance(FauxUnhashable(), Hashable)
Out[51]: True
```

Reference: https://docs.python.org/3/reference/datamodel.html#object.__hash__

samuelcolvin replied:
okay, so it can't just be a method on `BaseModel`, but let's make it a proper function, not a lambda.

layday replied:
> Instead of `.values` we should probably have `.items` here. The keys ought to be considered a part of the identity of the object.

I think this would unnecessary if we include a reference to the model class.

wozniakty commented on tests/test_main.py:
```diff
@@ -373,6 +386,32 @@ class Config:
+    m = TestModel()
+    with pytest.raises(TypeError) as exc_info:
+        hash(m)
+    assert "unhashable type: 'list'" in exc_info.value.args[0]
```
IMO would be nice for the exception message to be more like "unhashable type: 'TestModel' contains unhashable type: 'list'  but I wouldn't say it's a strong opinion.  Food for thought.

samuelcolvin reviewed (commented):
I think this is a good start, but my problem here is that I imagine in the following code `f` and `b` would have the same hash and therefore be "equal"

```py
class Foo(BaseModel):
    a: str
    b: int

    class Config:
        allow_mutation = False


class Bar(BaseModel):
    a: str
    b: int

    def dict(self, *args, **kwargs):
        d = super().dict(*args, **kwargs)
        d.pop('a')
        return d

    class Config:
        allow_mutation = False


f = Foo(a='xx', b=2)
b = Bar(a='xx', b=3)
```

Which is definitely not what people would expected.

I think we need to add some kind of reference to the qualname to the hash.

I think we also need to hide this behind another config option.

layday commented:
> I don't think this needs to be a lambda, it can be proper method on `BaseModel` I imagine. It could raise an error if `allow_mutation is True`.

The 'correct' way to make an object unhashable is by setting its `__hash__` attribute to `None`:

```
In [1]: class Foo:
   ...:     pass
   ...:

In [2]: class Bar:
   ...:     __hash__ = None
   ...:

In [3]: hash(Foo())
Out[3]: 284941123

In [4]: hash(Bar())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-d413494abdfb> in <module>
----> 1 hash(Bar())

TypeError: unhashable type: 'Bar'
```

This would have to happen on the metaclass and is done automatically by the Python interpreter for classes which implement `__eq__` but not `__hash__`:

```
In [5]: from pydantic import BaseModel

In [6]: class Baz(BaseModel):
   ...:     pass
   ...:

In [7]: hash(Baz())
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-67f7cb492b68> in <module>
----> 1 hash(Baz())

TypeError: unhashable type: 'Baz'

In [8]: Baz.__hash__

In [9]: class Bee:
   ...:     def __eq__(self, other):
   ...:         ...
   ...:

In [10]: Bee.__hash__

In [11]: Foo.__hash__
Out[11]: <slot wrapper '__hash__' of 'object' objects>
```

If `__hash__` were implemented on `BaseModel`, we'd have to emulate the default behaviour.  In addition,  if`__hash__` is not `None`, `isinstance(obj, typing.Hashable)` will not have the desired effect:

```
In [12]: from typing import Hashable

In [13]: isinstance(Bee(), Hashable)
Out[13]: False

In [14]: class FauxUnhashable:
    ...:     def __hash__(self):
    ...:         raise TypeError
    ...:

In [51]: isinstance(FauxUnhashable(), Hashable)
Out[51]: True
```

layday commented:
> I think we need to add some kind of reference to the qualname to the hash.

Would hashing the model class not work, e.g. `hash((self.__class__,) + tuple(self.__dict__.items()))`?  I know attrs combines the `__module__` name with the `__qualname__` but that might be because of something specific to attrs.

samuelcolvin commented:
I guess it might work, attrs might do it that way to support older versions of python.

samuelcolvin commented:
as long as this feature is hidden behind a 'hashable` config parameter, I'm not too bothered. We can say it's new and in "beta" or something in the docs.

layday commented:
hmm, would `hashable = True` imply `allow_mutation = False` or would they both need to be declared?  I don't think `hashable` is useful by itself or something that should be exposed - `hashable` without `allow_mutation` should be disallowed, in fact.  How about adding a new `frozen` setting which would (or could) supersede `allow_mutation` in 2.0?

rhuille commented:
So 2 things to modify:

- include the name of the class in the hash --> so we agree to diverge from the built in dataclass behavior ?

- add a boolean parameter `hashable` in BaseModel config. This parameter would have the following behavior:
   - if `hashable=True`: implement a hash function, 
            This hash function would raise an error if `allow mutation=True`. This is a way to "disallow hashable without allow_mutation" @layday  what do you think ?
            And to overcome this error, pydantic users should implement themself a hash function in the class model.
   - if `hashable=False` : the hash function is None.

layday commented:
* The class name is not unique.  We'd have to either combine the module name with the `__qualname__` or hash the class itself as suggested. 
* The hash function should not raise, a configuration error should be raised on model creation by the metaclass.  My preference is for `hashable` not being exposed.

samuelcolvin commented:
> How about adding a new `frozen` setting which would (or could) supersede allow_mutation in 2.0?

Sounds good to me.

> or hash the class itself as suggested

also sounds good to me.

samuelcolvin commented:
@rhuille are you going to finish this? otherwise let's lose it.

rhuille commented:
Hi @samuelcolvin yes I am doing this now. Thanks for the specifications.

rhuille commented:
Hi @samuelcolvin I have reworked the PR (and updated the description) to match the specifications specified by @layday and you. Thanks a lot for your feedback !

PrettyWood commented on changes/1880-rhuille.md:
```diff
@@ -0,0 +1 @@
+Add a new `frozen` boolean parameters. Setting `frozen=True` does everything that `allow_mutation=False` does, and also generate a hash function for the model i.e. `__hash__` is not `None`. This makes instances of the model potentially hashable if all the attributes are hashable. (default: `False`)
```
```suggestion
Add a new `frozen` boolean parameters. Setting `frozen=True` does everything that `allow_mutation=False` does, and also generates a `__hash__()` method for the model. This makes instances of the model potentially hashable if all the attributes are hashable. (default: `False`)
```

rhuille replied:
https://github.com/samuelcolvin/pydantic/pull/1881/commits/0d6eaec4bebb9e06fc2042f175a9c6fbf30deb18

PrettyWood commented on docs/usage/model_config.md:
```diff
@@ -26,6 +26,13 @@ Options:
+**`frozen`**
+
+!!! warning
+    This parameters is beta
```
```suggestion
    This parameter is in beta
```

rhuille replied:
https://github.com/samuelcolvin/pydantic/pull/1881/commits/0d6eaec4bebb9e06fc2042f175a9c6fbf30deb18

PrettyWood commented on docs/usage/model_config.md:
```diff
@@ -26,6 +26,13 @@ Options:
+!!! warning
+    This parameters is beta
+
+: setting `frozen=True` does everything that `allow_mutation=False` does, and also generate a hash function for the model i.e. `__hash__` is not `None`. This makes instances of the model potentially hashable if all the attributes are hashable. (default: `False`)
```
Same remark as change file

rhuille replied:
https://github.com/samuelcolvin/pydantic/pull/1881/commits/0d6eaec4bebb9e06fc2042f175a9c6fbf30deb18

PrettyWood commented on pydantic/main.py:
```diff
@@ -321,6 +329,7 @@ def __new__(mcs, name, bases, namespace, **kwargs):  # noqa C901
             '__custom_root_type__': _custom_root_type,
             '__private_attributes__': private_attributes,
             '__slots__': slots | private_attributes.keys(),
+            '__hash__': generate_hash_function(config.frozen, super()),
```
`super()` is the metaclass here. I don't understand why you need to do that 🤔

PrettyWood commented on pydantic/main.py:
```diff
@@ -198,6 +199,13 @@ def validate_custom_root_type(fields: Dict[str, ModelField]) -> None:
 
+def generate_hash_function(frozen: bool, super_class: Any) -> Optional[Callable[[Any], int]]:
+    def hash_function(self_: Any) -> int:
+        return hash(super_class) + hash(tuple(self_.__dict__.values()))
```
IMO you could just do `hash(self_.__class__)` instead of using `super_class`

rhuille replied:
Yes your are right ! https://github.com/samuelcolvin/pydantic/pull/1881/commits/12cff48de640035718d1308c04eb007fe5e1ef05 (I was overthinked this...)

PrettyWood commented on changes/1880-rhuille.md:
```diff
@@ -0,0 +1 @@
+Add a new `frozen` boolean parameters. Setting `frozen=True` does everything that `allow_mutation=False` does, and also generates a `__hash__()` method for the model. This makes instances of the model potentially hashable if all the attributes are hashable. (default: `False`)
```
```suggestion
Add a new `frozen` boolean parameter to `Config` (default: `False`).
Setting `frozen=True` does everything that `allow_mutation=False` does, and also generates a `__hash__()` method for the model. This makes instances of the model potentially hashable if all the attributes are hashable.
```

rhuille replied:
https://github.com/samuelcolvin/pydantic/pull/1881/commits/7397b678af3ef098d7e038a185e22f4edd973dfc

PrettyWood reviewed (commented):
Looks quite good.
Do we want to tackle the case `frozen=True` and `allow_mutation=True` and raise an error in this case ?

rhuille commented:
Good question ! For now, if `frozen=True` and `allow_mutation=True` the behavior is: "`frozen=True` is applied and `allow_mutation=True` is implicitly ignored". I guess I could add a warning message explaining that.

samuelcolvin commented on tests/mypy/modules/plugin_fail.py:
```diff
@@ -202,3 +202,37 @@ class AddProject:
+        frozen = True
+        extra = Extra.forbid
+
+        def config_method(self) -> None:
```
do you need this?

samuelcolvin commented on tests/mypy/modules/plugin_fail.py:
```diff
@@ -202,3 +202,37 @@ class AddProject:
+    x: int
+    y: str
+
+    def method(self) -> None:
```
can this be removed too?

samuelcolvin commented on tests/mypy/modules/plugin_fail.py:
```diff
@@ -202,3 +202,37 @@ class AddProject:
+            ...
+
+
+frozenmodel = FrozenModel(x=1, y='y', z='z')
```
this isn't testing frozen behaviour I think, again can be removed.

samuelcolvin commented on tests/mypy/modules/plugin_fail.py:
```diff
@@ -202,3 +202,37 @@ class AddProject:
+frozenmodel = FrozenModel(x=1, y='y', z='z')
+frozenmodel = FrozenModel(x=1)
+frozenmodel.y = 'a'
+FrozenModel.from_orm({})
+FrozenModel.from_orm({})  # type: ignore[pydantic-orm]  # noqa F821
```
again, are these lines testing frozen behaviour?

samuelcolvin commented on tests/mypy/modules/plugin_success.py:
```diff
@@ -139,3 +139,22 @@ class Model(BaseModel):
+
+
+class NotFrozenModel(FrozenModel):
+    a = 1
```
add a type here to avoid extra errors in `plugin-success-strict.txt `

samuelcolvin commented on tests/test_main.py:
```diff
@@ -351,39 +351,96 @@ class Model(BaseModel):
+
+    immutable = allow_mutation_ is False or frozen_ is True
+
+    if immutable:
```
rather than this if clause, better to setup two separate tests.

samuelcolvin reviewed (commented):
otherwise I think this looks great.

rhuille commented:
Hi @samuelcolvin , thanks for your review !

Since your review:

- I resolved the conflict with master: https://github.com/samuelcolvin/pydantic/pull/1881/commits/4815e4c42c3a4dd330da6c56c544f28c1af24897
- Following your comments:
    - I reworked the mypy tests so it test only the frozen behavior: https://github.com/samuelcolvin/pydantic/pull/1881/commits/7c3f4debf2ba2e6ad6c5f0659c2e1e89d74712a7 and 
    - I split the test function in 2: https://github.com/samuelcolvin/pydantic/pull/1881/commits/a6943a81bac2f0faf95bf33985eca3763808ab88 & https://github.com/samuelcolvin/pydantic/pull/1881/commits/3e98d572173c6ac46fc31a16298da2210a75fdd1 & https://github.com/samuelcolvin/pydantic/pull/1881/commits/ad82bf0484be65980cc6e9fdba2b5df1b91e7b38

Also, if you want me to, I can clean the commits history.


Thanks a lot, I am very proud of this work :D

samuelcolvin commented:
This is great, thank you so much.

davidhyman commented:
Hi, I've been playing with master (at 90df33cf1e01e0235e6069715a251e718bc07753) and am concerned that this eliminates the ability for a user-defined `__hash__`;

I'm all in favour of auto-generating it upon request, but I wonder if it would be less intrusive to user code to avoid setting `__hash__` in `frozen=False`? And even if it is `frozen=True`, perhaps a `__hash__` doesn't need generating if it's user-defined (less sure on that)?

propose changing
```
    '__hash__': generate_hash_function(config.frozen),
```
to
```
if config.frozen:
    new_namespace['__hash__'] = generate_hash_function()
```
(or some other mechanism e.g. popping again from the new namespace)

I might be missing something though - is there a suggested approach that would let me define the hash for my model? Without it, I can't put my model instances in a set, dictionary etc.

Motivating example: having a model which the `generate_hash_function` presently can't handle, such as nested complex generic types; or would be entirely unhashable in the usual sense due to mutable containers, but using a unique database id for reference - so as far as the application is concerned, the db ID is all that's needed for hashing?

PrettyWood commented:
I don't understand @davidhyman
For me the behaviour is right. When you write a custom `__hash__` it actually overwrites the default `__hash__` whether or not `frozen` is set.
Could you please share a code snippet of the problem?

davidhyman commented:
@PrettyWood I think you have fixed it in https://github.com/samuelcolvin/pydantic/pull/2423
and it was reported as
https://github.com/samuelcolvin/pydantic/issues/2422

Thank you! (and thanks to all the other contributors)

Sorry I was late to the party. I spent a bit of time on a repro today, originally thought I couldn't reproduce it and then realised that it's fixed (at least, in 1.8.1). I don't think I'd realised it was an inheritance thing; recalled commenting out the __hash__ assignment, so was sure it was something that had originally changed in this MR :thinking: .

```python
import pydantic

class P(pydantic.BaseModel):
    a: int

    def __hash__(self):
        """custom hash function"""
        return self.a

class M(P):
    """our hash function is still overwritten."""

    b: dict = dict()

    class Config:
        """pydantic config."""

        allow_mutation = False
        frozen = True

m = M(a=5)
assert m.__hash__() == 5
```
